procrastinate/worker.py

```diff
@@ -289,74 +213,188 @@ async def run_job(self, job: jobs.Job, worker_id: int) -> None:
+                extra=self._log_extra(
+                    action="finish_task", context=context, status=status
+                ),
+            )
 
-            text = (
-                f"Job {job.call_string} ended with status: {log_title}, "
-                f"lasted {duration:.3f} s"
+    async def _fetch_and_process_jobs(self):
+        """Fetch and process jobs until there is no job left or asked to stop"""
+        while not self._stop_event.is_set():
+            acquire_sem_task = asyncio.create_task(self._job_semaphore.acquire())
+            await utils.wait_any(acquire_sem_task, self._stop_event.wait())
+            if self._stop_event.is_set():
+                if acquire_sem_task.done():
+                    self._job_semaphore.release()
+                break
+            try:
+                job = await self.app.job_manager.fetch_job(queues=self.queues)
+            except BaseException:
+                self._job_semaphore.release()
+                raise
+
+            if not job:
+                self._notify_event.clear()
+                self._job_semaphore.release()
+                break
+
+            context = job_context.JobContext(
+                app=self.app,
+                worker_name=self.worker_name,
+                worker_queues=self.queues,
+                additional_context=self.additional_context.copy()
+                if self.additional_context
+                else {},
+                job=job,
+                task=self.app.tasks.get(job.task_name),
             )
-            if task_result:
-                text += f" - Result: {task_result}"[:250]
-            self.logger.log(log_level, text, extra=extra, exc_info=exc_info)
+            job_task = asyncio.create_task(self._process_job(context))
+            self._running_jobs[job_task] = context
 
-    def stop(self):
-        # Ensure worker will stop after finishing their task
-        self.stop_requested = True
-        # Ensure workers currently waiting are awakened
-        if self.notify_event:
-            self.notify_event.set()
+            def on_job_complete(task: asyncio.Task):
+                del self._running_jobs[task]
+                self._job_semaphore.release()
 
-        # Logging
+            job_task.add_done_callback(on_job_complete)
 
-        self.logger.info(
-            "Stop requested",
-            extra=self.base_context.log_extra(action="stopping_worker"),
-        )
+    async def run(self):
+        """
+        Run the worker
+        This will run forever until asked to stop/cancelled, or until no more job is available is configured not to wait
+        """
+        self.run_task = asyncio.current_task()
+        loop_task = asyncio.create_task(self._run_loop())
+
+        try:
+            # shield the loop task from cancellation
+            # instead, a stop signal is set to enable graceful shutdown
+            await asyncio.shield(loop_task)
+        except asyncio.CancelledError:
```

Comment:
I think we need to call [uncancel](https://docs.python.org/3/library/asyncio-task.html#asyncio.Task.uncancel):

> If end-user code is, for some reason, suppressing cancellation by catching [CancelledError](https://docs.python.org/3/library/asyncio-exceptions.html#asyncio.CancelledError), it needs to call this method to remove the cancellation state.

Reply:
That would be the thing to do if the cancellation was suppressed.

In this scenario, the run Task was cancelled and we re-raise the error after the graceful stop completes.

```python
worker_run_task = asyncio.run_task(worker.run())
worker.stop()
# this will complete successfully after graceful shutdown
await worker_run_task
```

```python
worker_run_task = asyncio.run_task(worker.run())
worker_run_task.cancel()
# this will raise CancelledError after graceful shutdown
await worker_run_task
```